README.md
# GORouter [![Build Status](https://travis-ci.org/illuminasy/gorouter.svg?branch=master)](https://travis-ci.org/illuminasy/gorouter) [![Coverage Status](https://coveralls.io/repos/github/illuminasy/gorouter/badge.svg?branch=master)](https://coveralls.io/github/illuminasy/gorouter?branch=master) [![GoDoc](https://godoc.org/github.com/illuminasy/gorouter?status.svg)](https://godoc.org/github.com/illuminasy/gorouter) [![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/illuminasy/gorouter/blob/master/LICENSE.md)

GoRouter a package which wraps around https://github.com/julienschmidt/httprouter
allows to start router with middlewares

Currently supported middlewares:

1) NewRelic - APM monitoring tool
2) Bugsnag - Log Management
 
## Usage

Get the library:

    $ go get -v github.com/illuminasy/gorouter

Just the router (no middleware)

```go
func startServer() {
    fmt.Println("Listening for http on port 80")
    router := gorouter.GetRouter(routes(), []string{})
    log.Fatal(http.ListenAndServe(":80", router))
}

func routes() gorouter.Routes {
    return gorouter.Routes{
        List: []gorouter.Route{
            gorouter.Route{
                Method: "GET",
                Path:   "/robots.txt",
                Handler: gorouter.PlainTextHandler([]string{
                    "User-agent: *",
                    "Disallow: /",
                }),
            },
            gorouter.Route{
                Method:  "GET",
                Path:    "/healthz",
                Handler: gorouter.JSONHandler(healthzHandler),
            },
            gorouter.Route{
                Method:  "GET",
                Path:    "/test",
                Handler: gorouter.HTMLHandler(healthzHandler),
            },
        },
        PanicHandler: func(w http.ResponseWriter, r *http.Request, err interface{}) {
            w.WriteHeader(http.StatusInternalServerError)
            w.Header().Set("Content-Type", "application/json")
            fmt.Fprint(w, "")
            log.Println(err)
        },
    }
}

func healthzHandler(w http.ResponseWriter, r *http.Request) (string, int) {
    return `{"status":"up"}`, http.StatusOK
}

```

Router with middlewares

```go
package somepackge

import "github.com/illuminasy/gorouter"
import "github.com/illuminasy/gorouter/middleware"

func startServer() {
    mc := middleware.Config{
        ErrorReportingConfig: middleware.ErrorReportingConfig{
            Enabled:      true,
            Bugsnag:      true,
            APIKey:       "testing",
            AppType:      "router",
            ReleaseStage: "Dev",
            AppVersion:   "0.1.0",
            ProjectPackages: []string{
                "main",
            },
            NotifyReleaseStages: []string{
                "Dev",
            },
            PanicHandler: func() {},
            Hostname: "localhost",
        },
        MetricCollectorConfig: middleware.MetricCollectorConfig{
            Enabled:  true,
            Newrelic: true,
            Debug:    true,
            AppName:  "TestApp",
            License:  "testing",
            Labels: map[string]string{
                "Environment": "Dev",
                "Version":     "0.1.0",
            },
            HostDisplayName: "localhost",
        },
    }

    fmt.Println("Listening for http on port 80")
    router := gorouter.GetRouterWithMiddleware(mc, routes(), []string{})
    log.Fatal(http.ListenAndServe(":80", router))
}

func routes() gorouter.Routes {
    return gorouter.Routes{
        List: []gorouter.Route{
            gorouter.Route{
                Method: "GET",
                Path:   "/robots.txt",
                Handler: gorouter.PlainTextHandler([]string{
                    "User-agent: *",
                    "Disallow: /",
                }),
            },
            gorouter.Route{
                Method:  "GET",
                Path:    "/healthz",
                Handler: gorouter.JSONHandler(healthzHandler),
            },
        },
        PanicHandler: func(w http.ResponseWriter, r *http.Request, err interface{}) {
            w.WriteHeader(http.StatusInternalServerError)
            w.Header().Set("Content-Type", "application/json")
            fmt.Fprint(w, "")
            log.Println(err)
        },
    }
}

func healthzHandler(w http.ResponseWriter, r *http.Request) (string, int) {
    return `{"status":"up"}`, http.StatusOK
}
```

Gorouter starts a web transaction by wrapping middleware around the handlers
but if you still need to create custom transactions (maybe non web stuff)

```go
    // w, r are w http.ResponseWriter, r *http.Request
    // these are optionals, just pass nil instead for non web transactions
    txn := middleware.GetMetricCollectorTransaction("someTxnID", "someTxnName", w, r)
    // Do something
    txn.End()
```

To create segments within a transaction

```go
    // w, r are w http.ResponseWriter, r *http.Request
    // these are optionals, just pass nil instead
    // if transaction doesnot exist then it creates new one.
    s := middleware.StartMetricCollectorSegment("someTxnID", "someTxnName", "someSegmentName", w, r)
    // Do something
    s.End()
```

To create datastore segments within a transaction say for mysql

```go
    dataStore := middleware.DataStore {
        Product: "mysql",
        Collection: "users",
        Operation: "INSERT",
        ParameterizedQuery: `INSERT INTO users (name, age) VALUES ($1, $2)"`,
        QueryParameters: map[string]interface{}{
            "name": "Dracula",
            "age": 439,
        },
        Host: "mysql-server-1",
        PortPathOrID: "3306",
        DatabaseName: "my_database",
    }
    // w, r are w http.ResponseWriter, r *http.Request
    // these are optionals, just pass nil instead
    // if transaction doesnot exist then it creates new one.
    s := middleware.StartMetricCollectorDataStoreSegment("someTxnID", "someTxnName", dataStore, w, r)
    // Do something
    s.End()
```

To report errors to metric collectors

```go
    // w, r are w http.ResponseWriter, r *http.Request
    // these are optionals, just pass nil instead
    // if transaction doesnot exist then it creates new one.
    middleware.MetricCollectorNoticeError("someTxnID", "someTxnName", errors.New("Invalid API config"), w, r)
```

Send Errors to bugsnag

```go
    middleware.ReportErrorToBugsnag("someErrorClass", errors.New("Invalid API config"))
```

## Thanks to

https://github.com/julienschmidt/httprouter

https://github.com/newrelic/go-agent

https://github.com/bugsnag/bugsnag-go


CHANGELOG.md
## 0.2.3 (October 15, 2022)
  - Added File Handler

## 0.2.2 (October 15, 2022)
  - Fixed request log nil error

## 0.2.1 (September 22, 2022)
  - Changed go version to v1.19

## 0.2.0 (September 21, 2022)
  - Upgraded packages

## 0.1.27 (May 26, 2019)
  - Added html handlers
  - Added static file handlers
  - Added ability to set status codes

## 0.1.26 (April 13, 2019)
  - Added Coverage badge to README.md

## 0.1.25 (April 13, 2019)
  - Fixed README.md

## 0.1.24 (April 13, 2019)
  - Fixed travis.yml

## 0.1.23 (April 13, 2019)
  - Fixed travis.yml

## 0.1.22 (April 13, 2019)
  - Fixed travis.yml

## 0.1.21 (April 13, 2019)
  - Fixed travis.yml

## 0.1.20 (April 13, 2019)
  - Fixed travis.yml

## 0.1.19 (April 13, 2019)
  - Fixed travis.yml

## 0.1.18 (April 13, 2019)
  - Fixed travis.yml

## 0.1.17 (April 13, 2019)
  - Fixed travis.yml

## 0.1.16 (April 13, 2019)
  - Fixed travis.yml

## 0.1.15 (April 13, 2019)
  - Fixed travis.yml

## 0.1.14 (April 13, 2019)
  - Updated README.md
  - Fixed travis.yml

## 0.1.13 (April 12, 2019)
  - Fixed bugsnag error reporting

## 0.1.12 (April 12, 2019)
  - Fixed segment reporting
  - Middleware Error Reporting will send errors to external such as bugsnag

## 0.1.11 (April 10, 2019)
  - Fixed undefined properties

## 0.1.10 (April 10, 2019)
  - Fixed import issues_2

## 0.1.9 (April 10, 2019)
  - Generic metric collector and error reporting

## 0.1.8 (April 09, 2019)
  - Fixed import issues

## 0.1.7 (April 09, 2019)
  - Made Metric Collectore more generic

## 0.1.6 (April 06, 2019)
  - Updated dependencies

## 0.1.5 (April 06, 2019)
  - Fixed import issues

## 0.1.4 (April 06, 2019)
  - HTTP Writer and response can be accessed by the handlers

## 0.1.3 (January 13, 2019)
  - Fixed Middleware imports.

## 0.1.2 (January 13, 2019)


## 0.1.1 (January 13, 2019)
  - Fixed ReadMe file.

## 0.1.0 (January 12, 2019)
  - Wrapper for httpRouter
  - Added middleware support for newrelic
  - Added middleware support for bugsnag



gorouter_test.go
package gorouter

import (
	"fmt"
	"log"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/illuminasy/gorouter/middleware"
	"github.com/stretchr/testify/assert"
)

var routes = Routes{
	List: []Route{
		Route{
			Method: "GET",
			Path:   "/robots.txt",
			Handler: PlainTextHandler([]string{
				"User-agent: *",
				"Disallow: /",
			}),
		},
		Route{
			Method: "GET",
			Path:   "/healthz",
			Handler: JSONHandler(func(w http.ResponseWriter, r *http.Request) (string, int) {
				return `{"status":"up"}`, http.StatusOK
			}),
		},
		Route{
			Method: "GET",
			Path:   "/panic",
			Handler: JSONHandler(func(w http.ResponseWriter, r *http.Request) (string, int) {
				panic("testing...")
			}),
		},
	},
	PanicHandler: func(w http.ResponseWriter, r *http.Request, err interface{}) {
		w.WriteHeader(http.StatusInternalServerError)
		w.Header().Set("Content-Type", "application/json")
		fmt.Fprint(w, "")
		log.Println(err)
	},
}

var mc = middleware.Config{
	ErrorReportingConfig: middleware.ErrorReportingConfig{
		Enabled:      true,
		Bugsnag:      true,
		APIKey:       "testing",
		ReleaseStage: "Testing",
		ProjectPackages: []string{
			"main",
		},
		NotifyReleaseStages: []string{
			"Testing",
		},
		AppVersion: "0.1.0",
	},
	MetricCollectorConfig: middleware.MetricCollectorConfig{
		Enabled:  true,
		Newrelic: true,
		Debug:    true,
		AppName:  "Testing",
		License:  "testing",
		Labels: map[string]string{
			"Environment": "Dev",
			"Version":     "0.1.0",
		},
		HostDisplayName: "testing.localhost",
	},
}

func TestGetRobots(t *testing.T) {
	respRec := httptest.NewRecorder()
	req, _ := http.NewRequest("GET", "/robots.txt", nil)

	router := GetRouter(routes, []string{})

	router.ServeHTTP(respRec, req)
	assert.Equal(t, http.StatusOK, respRec.Code)
	assert.Equal(t, "User-agent: *\nDisallow: /", string(respRec.Body.Bytes()))
}

func TestGetHealthz(t *testing.T) {
	respRec := httptest.NewRecorder()
	req, _ := http.NewRequest("GET", "/healthz", nil)

	router := GetRouter(routes, []string{})

	router.ServeHTTP(respRec, req)
	assert.Equal(t, http.StatusOK, respRec.Code)
	assert.Contains(t, string(respRec.Body.Bytes()), `{"status":"up"}`)
}

func TestNonExistent(t *testing.T) {
	respRec := httptest.NewRecorder()
	req, _ := http.NewRequest("GET", "/non-existent-endpoint", nil)

	router := GetRouter(routes, []string{})

	router.ServeHTTP(respRec, req)
	assert.Equal(t, http.StatusNotFound, respRec.Code)
	assert.Equal(t, string(respRec.Body.Bytes()), "404 page not found\n")
}

func TestPanic(t *testing.T) {
	respRec := httptest.NewRecorder()
	req, _ := http.NewRequest("GET", "/panic", nil)

	router := GetRouter(routes, []string{})

	router.ServeHTTP(respRec, req)
	assert.Equal(t, http.StatusInternalServerError, respRec.Code)
}

func TestGetRobotsWithMiddleware(t *testing.T) {
	respRec := httptest.NewRecorder()
	req, _ := http.NewRequest("GET", "/robots.txt", nil)

	router := GetRouterWithMiddleware(mc, routes, []string{})

	router.ServeHTTP(respRec, req)
	assert.Equal(t, http.StatusOK, respRec.Code)
	assert.Equal(t, "User-agent: *\nDisallow: /", string(respRec.Body.Bytes()))
}

func TestGetHealthzWithMiddleware(t *testing.T) {
	respRec := httptest.NewRecorder()
	req, _ := http.NewRequest("GET", "/healthz", nil)

	router := GetRouterWithMiddleware(mc, routes, []string{})

	router.ServeHTTP(respRec, req)
	assert.Equal(t, http.StatusOK, respRec.Code)
	assert.Contains(t, string(respRec.Body.Bytes()), `{"status":"up"}`)
}

func TestNonExistentWithMiddleware(t *testing.T) {
	respRec := httptest.NewRecorder()
	req, _ := http.NewRequest("GET", "/non-existent-endpoint", nil)

	router := GetRouterWithMiddleware(mc, routes, []string{})

	router.ServeHTTP(respRec, req)
	assert.Equal(t, http.StatusNotFound, respRec.Code)
	assert.Equal(t, string(respRec.Body.Bytes()), "404 page not found\n")
}

func TestPanicWithMiddleware(t *testing.T) {
	respRec := httptest.NewRecorder()
	req, _ := http.NewRequest("GET", "/panic", nil)

	router := GetRouterWithMiddleware(mc, routes, []string{})

	router.ServeHTTP(respRec, req)
	assert.Equal(t, http.StatusInternalServerError, respRec.Code)
}


gorouter.go
package gorouter

import (
	"fmt"
	"net/http"
	"strings"

	"github.com/illuminasy/gorouter/middleware"
	"github.com/julienschmidt/httprouter"
)

// Routes List of routes and panic handler
type Routes struct {
	List         []Route
	PanicHandler func(w http.ResponseWriter, r *http.Request, err interface{})
}

// Route Each route needs url path, type of request and a handler
type Route struct {
	Method  string
	Path    string
	Handler httprouter.Handle
}

var headersAllowedByCORS = []string{
	"Host",
	"Content-Type",
	"Connection",
	"User-Agent",
	"Cache-Control",
	"Accept-Encoding",
}

// PlainTextHandler handles plain text responses with appropriate headers
func PlainTextHandler(lines []string) httprouter.Handle {
	return func(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
		contents := []byte(strings.Join(lines, "\n"))
		w.WriteHeader(http.StatusOK)
		w.Header().Set("Content-Type", "text/plain")
		w.Write(contents)
	}
}

// JSONHandler handles json responses with appropriate headers
func JSONHandler(handler func(http.ResponseWriter, *http.Request) (string, int)) httprouter.Handle {
	return func(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
		body, statusCode := handler(w, r)
		w.WriteHeader(statusCode)
		w.Header().Set("Content-Type", "application/json")
		w.Write([]byte(body))
	}
}

// HTMLHandler handles html responses with appropriate headers
func HTMLHandler(handler func(http.ResponseWriter, *http.Request) (string, int)) httprouter.Handle {
	return func(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
		body, statusCode := handler(w, r)
		w.WriteHeader(statusCode)
		w.Header().Set("Content-Type", "text/html; charset=utf-8")
		w.Write([]byte(body))
	}
}

// FileHandler handles files responses with appropriate headers
func FileHandler(handler func(http.ResponseWriter, *http.Request) (string, int)) httprouter.Handle {
	return func(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
		path, statusCode := handler(w, r)
		w.WriteHeader(statusCode)
		http.FileServer(http.Dir(path))
	}
}

// StaticFileHandler handles static files responses with appropriate headers
func StaticFileHandler(path string) httprouter.Handle {
	return func(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
		w.WriteHeader(http.StatusOK)
		http.FileServer(http.Dir(path))
	}
}

// GetRouter returns a router, optionally additional headers can be passed to set
func GetRouter(routes Routes, additionalHeaders []string) *httprouter.Router {
	return createRouter(routes, additionalHeaders, middleware.Config{})
}

// GetRouterWithMiddleware returns a router with middlewares wrapped around it, optionally additional headers can be passed to set
func GetRouterWithMiddleware(mc middleware.Config, routes Routes, additionalHeaders []string) http.Handler {
	router := createRouter(routes, additionalHeaders, mc)
	return middleware.ErrorReportingMiddleware(router, mc.ErrorReportingConfig)
}

// createRouter creates a router
func createRouter(routes Routes, additionalHeaders []string, mc middleware.Config) *httprouter.Router {
	apiMethods := map[string][]string{}

	router := httprouter.New()
	router.PanicHandler = routes.PanicHandler
	for _, route := range routes.List {
		apiMethods[route.Path] = append(apiMethods[route.Path], route.Method)
		router.Handle(route.Method, route.Path, wrapMiddlewares(route.Handler, route.Path, mc))
	}

	for k, v := range apiMethods {
		router.OPTIONS(k, constructOptions(v, additionalHeaders))
	}

	return router
}

func wrapMiddlewares(handler httprouter.Handle, path string, mc middleware.Config) httprouter.Handle {
	return middleware.Wrapper(handler, path, mc)
}

func constructOptions(methods []string, additionalHeaders []string) func(http.ResponseWriter, *http.Request, httprouter.Params) {
	methodCsv := strings.Join(append(methods, "OPTIONS"), ",")
	return func(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
		w.Header().Set("Access-Control-Allow-Origin", "*")
		decorateWithCORS(w.Header(), methodCsv, additionalHeaders)
		w.Header().Set("Content-Type", "application/json")
		fmt.Fprint(w, "{}")
	}
}

func decorateWithCORS(headers http.Header, methods string, additionalHeaders []string) {
	allowedHeaders := append(headersAllowedByCORS, additionalHeaders...)
	headers.Set("Access-Control-Allow-Origin", "*")
	headers.Set("Access-Control-Allow-Methods", methods)
	headers.Set("Access-Control-Allow-Headers",
		strings.Join(allowedHeaders, ","),
	)
}


middleware/bugsnag.go
package middleware

import (
	"fmt"
	"net/http"

	bugsnag "github.com/bugsnag/bugsnag-go"
)

var bugsnagConfigured = false

func configureBugsnag(config ErrorReportingConfig) {
	bugsnag.Configure(bugsnag.Configuration{
		APIKey:              config.APIKey,
		ReleaseStage:        config.ReleaseStage,
		AppType:             config.AppType,
		AppVersion:          config.AppVersion,
		ProjectPackages:     config.ProjectPackages,
		NotifyReleaseStages: config.NotifyReleaseStages,
		ParamsFilters:       []string{"password", "secret"},
		PanicHandler:        func() {},
		Hostname:            config.Hostname,
	})

	bugsnagConfigured = true
}

// bugsnagMiddleware configures and wraps bugsnag handler around the router
func bugsnagMiddleware(router http.Handler, config ErrorReportingConfig) http.Handler {
	configureBugsnag(config)
	return bugsnag.Handler(router)
}

func sendError(errorClass string, err error, a ...interface{}) error {
	if bugsnagConfigured {
		// append error class so bugsnag can group errors using this
		a = append([]interface{}{bugsnag.ErrorClass{Name: errorClass}}, a...)
		return bugsnag.Notify(err, a...)
	}

	return fmt.Errorf("Bugsnag has not been configured")
}


middleware/error_reporting.go
package middleware

import (
	"net/http"
)

// ErrorReportingConfig Error reporting config
type ErrorReportingConfig struct {
	Enabled             bool
	Bugsnag             bool
	APIKey              string
	ReleaseStage        string
	AppType             string
	AppVersion          string
	ProjectPackages     []string
	NotifyReleaseStages []string
	ParamsFilters       []string
	PanicHandler        func()
	Hostname            string
}

// ErrorReportingMiddleware configures and wraps error reporting handler around the router
func ErrorReportingMiddleware(router http.Handler, config ErrorReportingConfig) http.Handler {
	if config.Enabled && config.Bugsnag {
		return bugsnagMiddleware(router, config)
	}

	return router
}

// ReportErrorToBugsnag send error to bugsnag
func ReportErrorToBugsnag(errorClass string, err error, a ...interface{}) error {
	return sendError(errorClass, err, a...)
}


middleware/metric_collector.go
package middleware

import (
	"net/http"

	"github.com/julienschmidt/httprouter"
	newrelic "github.com/newrelic/go-agent"
)

// MetricCollectorConfig Metric Collector config
type MetricCollectorConfig struct {
	Enabled         bool
	Newrelic        bool
	Debug           bool
	AppName         string
	License         string
	Labels          map[string]string
	HostDisplayName string
}

// MetricCollectorTxn Metric Collector transaction
type MetricCollectorTxn struct {
	Txn newrelic.Transaction
}

// MetricCollectorSegment Metric Collector segment
type MetricCollectorSegment struct {
	Segment *newrelic.Segment
}

// MetricCollectorDatastoreSegment Metric Collector datastore segment
type MetricCollectorDatastoreSegment struct {
	DatastoreSegment newrelic.DatastoreSegment
}

// DataStore Metric Collector datastore
type DataStore struct {
	Product            string
	Collection         string
	Operation          string
	ParameterizedQuery string
	QueryParameters    map[string]interface{}
	Host               string
	PortPathOrID       string
	DatabaseName       string
}

// metricCollectorMiddleware Creates and starts Metric Collector middlware
func metricCollectorMiddleware(handler httprouter.Handle, path string, config MetricCollectorConfig) httprouter.Handle {
	if config.Newrelic && len(config.License) > 0 {
		return newrelicMiddleware(handler, path, config)
	}

	return handler
}

// GetMetricCollectorTransaction returns new or existing metric collector transanction
func GetMetricCollectorTransaction(txnID string, txnName string, w http.ResponseWriter, r *http.Request) MetricCollectorTxn {
	var mcTxn MetricCollectorTxn
	if newRelicApp != nil {
		mcTxn.Txn = getNewrelicTransaction(txnID, txnName, w, r)
	}

	return mcTxn
}

// StartMetricCollectorSegment Starts and retuns a metric collector segment for a transaction
func StartMetricCollectorSegment(txnID string, txnName string, segmentName string, w http.ResponseWriter, r *http.Request) MetricCollectorSegment {
	var mxnSgmt MetricCollectorSegment

	if newRelicApp != nil {
		mxnSgmt.Segment = startNewrelicSegment(txnID, txnName, segmentName, w, r)
	}

	return mxnSgmt
}

// StartMetricCollectorDataStoreSegment Starts and retuns a metric collector datastore segment for a transaction
func StartMetricCollectorDataStoreSegment(txnID string, txnName string, datastore DataStore, w http.ResponseWriter, r *http.Request) MetricCollectorDatastoreSegment {
	var mxnDsSgmt MetricCollectorDatastoreSegment

	if newRelicApp != nil {
		mxnDsSgmt.DatastoreSegment = startNewrelicDataStoreSegment(txnID, txnName, datastore, w, r)
	}

	return mxnDsSgmt
}

// MetricCollectorNoticeError Send error to newrelic
func MetricCollectorNoticeError(txnID string, txnName string, err error, w http.ResponseWriter, r *http.Request) {
	if newRelicApp != nil {
		newrelicNoticeError(txnID, txnName, err, w, r)
	}
}


middleware/newrelic.go
package middleware

import (
	"log"
	"net/http"
	"os"

	"github.com/julienschmidt/httprouter"
	newrelic "github.com/newrelic/go-agent"
)

var newRelicApp newrelic.Application

var newrelicTransactionList map[string]map[string]newrelic.Transaction

type newrelicDataStore struct {
	Product            string
	Collection         string
	Operation          string
	ParameterizedQuery string
	QueryParameters    map[string]interface{}
	Host               string
	PortPathOrID       string
	DatabaseName       string
}

func newrelicMiddleware(handler httprouter.Handle, path string, config MetricCollectorConfig) httprouter.Handle {
	return func(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
		newRelicApp, err := getnewrelicApp(config)
		if err != nil {
			log.Println(err)
		}

		if newRelicApp != nil && err == nil {
			txn := newRelicApp.StartTransaction(path, w, r)
			defer func() {
				err := txn.End()
				if err != nil {
					log.Println(err)
				}
			}()
			handler(txn, r, ps)
			return
		}

		handler(w, r, ps)
	}
}

func getnewrelicApp(config MetricCollectorConfig) (newrelic.Application, error) {
	var err error
	if newRelicApp == nil {
		err = configureNewRelic(config)
	}

	return newRelicApp, err
}

func configureNewRelic(config MetricCollectorConfig) error {
	cfg := newrelic.NewConfig(config.AppName, config.License)
	cfg.Enabled = config.Enabled
	cfg.Labels = config.Labels
	cfg.HostDisplayName = config.HostDisplayName
	cfg.DistributedTracer.Enabled = true

	if config.Debug {
		cfg.Logger = newrelic.NewDebugLogger(os.Stdout)
	} else {
		cfg.Logger = newrelic.NewLogger(os.Stdout)
	}

	err := cfg.Validate()
	if err != nil {
		return err
	}

	newRelicApp, err = newrelic.NewApplication(cfg)
	return err
}

func getNewrelicTransaction(id string, name string, w http.ResponseWriter, r *http.Request) newrelic.Transaction {
	if len(newrelicTransactionList) == 0 {
		newrelicTransactionList = make(map[string]map[string]newrelic.Transaction)
	}

	if _, ok := newrelicTransactionList[id]; !ok {
		newrelicTransactionList[id] = make(map[string]newrelic.Transaction)
	}

	if txn, ok := w.(newrelic.Transaction); ok {
		newrelicTransactionList[id][name] = txn
		return txn
	}

	if list, ok := newrelicTransactionList[id]; ok {
		if txn, ok := list[name]; ok {
			return txn
		}
	}

	var txn newrelic.Transaction

	if newRelicApp != nil {
		txn = newRelicApp.StartTransaction(name, w, r)
		newrelicTransactionList[id][name] = txn
	}

	return txn
}

func startNewrelicSegment(txnID string, txnName string, segmentName string, w http.ResponseWriter, r *http.Request) *newrelic.Segment {
	txn := getNewrelicTransaction(txnID, txnName, w, r)
	segment := newrelic.StartSegment(txn, segmentName)

	return segment
}

func startNewrelicDataStoreSegment(txnID string, txnName string, datastore DataStore, w http.ResponseWriter, r *http.Request) newrelic.DatastoreSegment {
	txn := getNewrelicTransaction(txnID, txnName, w, r)
	s := newrelic.DatastoreSegment{
		Product:            newrelic.DatastoreMySQL,
		Collection:         "users",
		Operation:          "INSERT",
		ParameterizedQuery: "INSERT INTO users (name, age) VALUES ($1, $2)",
		QueryParameters: map[string]interface{}{
			"name": "Dracula",
			"age":  439,
		},
		Host:         "mysql-server-1",
		PortPathOrID: "3306",
		DatabaseName: "my_database",
	}
	s.StartTime = newrelic.StartSegmentNow(txn)

	return s
}

func newrelicNoticeError(txnID string, txnName string, err error, w http.ResponseWriter, r *http.Request) {
	txn := getNewrelicTransaction(txnID, txnName, w, r)
	txn.NoticeError(err)
}
